src/charm.py

```diff
@@ -39,22 +65,43 @@ def _on_start(self, event):
         sssd.start()
         self.unit.status = ActiveStatus("SSSD Operator Started")
 
-    def _on_sssd_ldap_relation_changed(self, event):
-        """Handle sssd-ldap relation changed event."""
-        auth_relation = self.model.get_relation("sssd-ldap")
-        ca_cert = auth_relation.data[event.app].get("ca-cert")
-        sssd_conf = auth_relation.data[event.app].get("sssd-conf")
-        if None not in [ca_cert, sssd_conf]:
-            try:
-                sssd.save_ca_cert(ca_cert)
-            except Exception:
-                self.unit.status = BlockedStatus("CA Certificate transfer failed")
-            sssd.save_conf(sssd_conf)
-            logger.debug("sssd-ldap relation-changed data found.")
-        else:
-            logger.error("sssd-ldap relation-changed data not found: ca-cert and sssd-conf.")
+    def _on_certificate_unavailable(self, event: CertificateUnavailableEvent):
+        """Handle certificate-unavailable event."""
+        self.unit.status = BlockedStatus("CA Certificate not available")
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent):
+        """Handle certificate-available event."""
+        try:
+            sssd.save_ca_cert(event.ca_cert)
+        except Exception:
+            self.unit.status = BlockedStatus("CA Certificate secret transfer failed")
+
+    def _on_config_data_available(self, event: ConfigDataAvailableEvent):
+        """Handle certificate-available event."""
+        sssd.save_conf(
+            event.basedn,
+            self.app.name,
+            event.ldap_uri,
+            event.ldbd_content,
+            event.lp_content,
+        )
+        self.unit.status = ActiveStatus("SSSD Active")
+
+    def _on_ldap_ready(self, event: LdapReadyEvent):
+        """Handle ldap-ready event."""
+        sssd.restart()
         if not sssd.running:
             logger.error("Failed to start sssd")
+            self.unit.status = BlockedStatus("SSSD failed to run")
+
+    def _on_server_unavailable(self, event: ServerUnavailableEvent):
+        """Handle server-unavailable event."""
+        # Stop SSSD
+        sssd.stop()
+        # Remove Obsolete SSSD Configuration
+        sssd.remove_conf()
+        # Remove Obsolete CA Certificate
+        sssd.remove_ca_cert()
```

Comment:
I don't think you need the comments here. Your function names are clear so I can easily gather what these methods are doing. If I want more details, I can read the docstrings!

```suggestion
        sssd.stop()
        sssd.remove_conf()
        sssd.remove_ca_cert()
```